packages/aws-cdk/lib/import.ts

```diff
@@ -0,0 +1,53 @@
+import * as cfnDiff from '@aws-cdk/cloudformation-diff';
+import * as cxapi from '@aws-cdk/cx-api';
+import * as promptly from 'promptly';
+import { ResourcesToImport } from './api/util/cloudformation';
+
+// Basic idea: we want to have a structure (ideally auto-generated from CFN definitions) that lists all resource types
+// that support importing and for each type, the identification information
+//
+// For each resource that is to be added in the new template:
+// - look up the identification information for the resource type [if not found, fail "type not supported"]
+// - look up the physical resource (perhaps using cloud control API?) [if not found, fail "resource to be imported does not exist"]
+// - assembe and return "resources to import" object to be passed on to changeset creation
+//
+// TEST: can we have a CFN changeset that both creates resources and import other resources?
+const RESOURCE_IDENTIFIERS: { [key: string]: string[] } = {
+  'AWS::S3::Bucket': ['BucketName'],
+};
+
+export async function prepareResourcesToImport(oldTemplate: any, newTemplate: cxapi.CloudFormationStackArtifact): Promise<ResourcesToImport> {
+  const diff = cfnDiff.diffTemplate(oldTemplate, newTemplate.template);
+
+  const additions: { [key: string]: cfnDiff.ResourceDifference } = {};
+  diff.resources.forEachDifference((id, chg) => {
+    if (chg.isAddition) {
+      additions[id] = chg;
+    }
+  });
+
+  const resourcesToImport: ResourcesToImport = [];
+  for (let [id, chg] of Object.entries(additions)) {
+    if (chg.newResourceType === undefined || !(chg.newResourceType in RESOURCE_IDENTIFIERS)) {
+      throw new Error(`Resource ${id} is of type ${chg.newResourceType} that is not supported for import`);
+    }
+
+    let identifier: { [key: string]: string } = {};
+    for (let idpart of RESOURCE_IDENTIFIERS[chg.newResourceType]) {
+      if (chg.newProperties && (idpart in chg.newProperties)) {
+        identifier[idpart] = chg.newProperties[idpart];
+      } else {
+        const displayName : string = newTemplate.template?.Resources?.[id]?.Metadata?.['aws:cdk:path'] ?? id;
+        identifier[idpart] = await promptly.prompt(`Please enter ${idpart} of ${chg.newResourceType} to import as ${displayName.replace(/\/Resource$/, '')}: `);
+      }
+    }
+
+    resourcesToImport.push({
+      LogicalResourceId: id,
+      ResourceType: chg.newResourceType,
+      ResourceIdentifier: identifier,
```

Comment:
Oh god must we really always supply this, even if it's already in the template? Great job on automating that away!

Reply:
Yeah, the CFN API for importing is suboptimal ... :/ This structure can at least be automatically assembled, but what's worse, a changeset of type import cannot contain any non-import template changes apparently. It will be a challenge to support an import of higher-level constructs that synthesise into multiple CFN resources of which only a subset must be imported and the rest needs to be created.

Reply:
> It will be a challenge to support an import of higher-level constructs that synthesise into multiple CFN resources of which only a subset must be imported and the rest needs to be created.

That is a good point. What we *could* be doing is:

* From all the resources that are added, have the user pick which ones they want to import (exact UI for this TBD)
* Cull the template to only selected resources
* Run an import with only those resources
* Follow up with a regular deploy

By necessity the resources that are imported cannot depend on other resources that aren't being imported as well (or exist already) so that's a bit limiting but best we can do for now until we build a more comprehensive importing engine?

Reply:
That sounds like a great idea, I will try to implement that!

Reply:
As for the UI, it might be as simple as: 

* For resources we don't know a name for, if the user doesn't fill in a name we skip it.
* For sources we DO know a name for, we pre-fill it and the user can backspace it if they don't want to import. 

The latter may not be possible with `promptly`, I'm not quite sure. But that would seem simple and clear enough.

Reply:
@rix0rrr , I have implemented the above flow: pre-filling a prompt wasn't possible (or I didn't find out how) but it now works that:
* if the identifier can be inferred from props, only ask whether to import the given resource or skip
* if the identifier can't be inferred, ask for the identifier interactively, if left empty, skip import

All resources that don't support importing or were marked to be skipped by user will be removed from the template for the import operation. So the import succeeds and the subsequent deploy will create the remaining resources.

What do you think?

Reply:
That sounds great!